### PR TITLE
Template DataVector to DataVectorImpl<type>

### PR DESCRIPTION
**Pull request by moxcodes**
Converts existing DataVector class construction to a templated DataVectorImpl<type>.
Added `using DataVector = DataVectorImpl<double>` to ensure all existing code remains
functional. All functions that survived templatization are templated, while the rest are
left as template specializations. Some shuffling between DataVector.hpp and DataVector.cpp
as necessary for inlines etc.

All forward declarations `class DataVector;` are replaced with

`template <typename T>
class DataVectorImpl;
using DataVector = DataVectorImpl<double>`

### Types of changes:

- [x] New feature

### Component:

- [x] Code

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`. For
  instructions on how to perform the CI checks locally refer to the [Dev guide
  on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run `make doc`
  to generate the documentation locally into `BUILD_DIR/docs/html`. Then open
  `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

File: src/ApparentHorizons/Tags.hpp [hunk 1]
```diff
@@ -12,7 +12,9 @@
 #include "Utilities/ForceInline.hpp"
 #include "Utilities/TMPL.hpp"
 
-class DataVector;
+template <typename T>
+class DataVectorImpl;
+using DataVector = DataVectorImpl<double>;
```
Comment (nilsdeppe) [inline src/ApparentHorizons/Tags.hpp:17]: We discussed using a strong typedef rather than a type alias so you don't need to do this everywhere. Then you can forward declare the strong type def. That is,

```
struct DataVector: public DataVectorImpl<double> {
  using DataVectorImpl<double>::DataVectorImpl;
  // Any other functions that should become visible
};
```

Comment (markscheel) [reply src/ApparentHorizons/Tags.hpp:17]: Ok, I like this.  Not sure if any other functions need to be visible inside `struct DataVector` other than the constructor.

Comment (moxcodes) [reply src/ApparentHorizons/Tags.hpp:17]: It looks like there are a couple of things that need to be declared in the `struct DataVector` for this, but I think I'll be able to keep it pretty minimal. I'll revise with the strong typedef version.

Comment (moxcodes) [reply src/ApparentHorizons/Tags.hpp:17]: It appears to be true that if we use inheritance rather than the transparent typedef, we will also need to re-implement potentially templatable operators, limiting the usefulness of the templating structure (particularly e.g. `==` acting on arrays of `DataVector`s -- there are several such implementations in `DataVector.hpp`). It is my understanding that the compiler will not directly use the operators associated with the base class in this case (when the called arguments are `array<Derived, dim>`, strong typedef ensures that it will not template deduce to a declared  `array<Base<spec>,dim>`, even when `Derived` inherits from `Base<spec>`).
Does this change any perspectives on strong typedef vs `using` declaration?
short version: do we want annoying forward DataVector declares, or code duplication for operators on classes templated on DataVectors of different types?

Comment (nilsdeppe) [reply src/ApparentHorizons/Tags.hpp:17]: As per my last comment on https://github.com/sxs-collaboration/spectre/pull/715, "This whole PR seems largely copy-pasta. I think we should implement most of this stuff in a macro and then use the code in both DataVector and ModalVector." I think that will make it easier for the various vector structures we want. This will also mean we don't have to worry about users getting errors pointing to some `Impl` they have to debug.

Comment (moxcodes) [reply src/ApparentHorizons/Tags.hpp:17]: Okay, I think I like using macros for these -- I'll put that in.
